Add support for Redis and Redis Cluster

qooba opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. Ensure that your code follows our code conventions: https://github.com/feast-dev/feast/blob/master/docs/contributing.md#code-conventions
2. Run unit tests and ensure that they are passing: https://github.com/feast-dev/feast/blob/master/docs/contributing.md#running-unit-tests
3. If your change introduces any API changes, make sure to update the integration tests scripts here: https://github.com/feast-dev/feast/tree/master/tests/e2e
4. Make sure documentation is updated for your PR!
5. Make sure you have signed the CLA https://cla.developers.google.com/clas

-->

**What this PR does / why we need it**:
PR adds Redis and RedisCluster support as online store. As described in https://github.com/feast-dev/feast/issues/1497 it is backward compatible with Feast 0.9 thus it can be used for migration to 0.10

**Which issue(s) this PR fixes**:
Fixes #1497

**Does this PR introduce a user-facing change?**:
```release-note
Redis and RedisCluster online stores support added. Format is compatible with Feast 0.9.
```

woop commented:
/ok-to-test

woop commented:
This is great work @qooba! Really surprised to see the implementation so quickly. We're still in the process of collecting feedback on the #1497, given that we may need to support a more scalable ingestion API for real time events. We may need a little bit of time before we can review and merge this code in, but we're super thankful for the contribution!!

woop commented:
One thing that stands out is that we don't have any automated tests for Redis as part of this PR. We'll definitely need to extend the PR to also include an integration test. @qooba do you want to take a stab at that?

qooba commented:
Sure @woop I will extend PR with integration tests.

qooba commented:
/retest

woop commented:
For some reason it's not allowing you to run tests until you've actually merged some code into Feast. I'll need to approve every time.

jklegar reviewed (commented):
Thanks for the PR @qooba! For integration tests, adding a `sdk/python/tests/test_cli_redis.py` that is the equivalent of `sdk/python/tests/test_cli_gcp.py` should cover the main parts, plus adding redis into the GitHub actions workflows like Willem said above

qooba commented:
@jklegar I have added ```sdk/python/tests/test_cli_redis.py``` and also ```sdk/python/tests/test_offline_online_store_consistency.py::test_redis_offline_online_store_consistency``` to check offline consistency with redis. Redis service also added to GitHub action workflow.

qooba commented:
The test ```sdk/python/tests/test_cli_redis.py::test_basic``` fails because added configuration is not used:
```
env:
  REDIS_TYPE: REDIS
  REDIS_CONNECTION_STRING: redis:6379,db=0
```

@jklegar can you help with this ?

jklegar commented:
@qooba I've added the redis service to pr_integration_tests.yml in master, can you rebase this diff? Once you rebase it'll run the updated action here

qooba commented:
@jklegar there will be small change in github action configuration:
https://github.com/feast-dev/feast/pull/1511/commits/b8b0d9f39d81c9a97ac2d8f538d884250115ea1c

Additionally **macOS-latest** doesn't have docker thus tests will never pass can you remove it :) 

And commit once again to master then I will rebase.

jklegar commented:
Thanks @qooba and apologies for the delay here, it should be good to rebase now

qooba commented:
Thanks @jklegar rebased and tests are passing now :)

woop commented:
/kind feature

abrookins commented:
Hi, Andrew from Redis Labs here! I'll give this a read-through today. :)

woop commented on sdk/python/feast/infra/redis_provider.py:
```diff
@@ -0,0 +1,276 @@
+        """
+        startup_nodes, kwargs = self._get_cs()
+
+        if os.environ["REDIS_TYPE"] == "REDIS_CLUSTER":
```
We certainly should be moving to using purely `feature_store.yaml` for configuration, and not environmental variables. It will be too hard to keep track of these configuration backdoors.

abrookins replied:
Does `feature_store.yaml` support provider-specific configuration? If it does, I agree that using it may be better than environment variables, except the connection string, of course. I assume the project YAML is supposed to be stored with the project in git, etc., so you wouldn't add connection details there?

woop replied:
Yes, that is the goal. Also, I think we can add environmental variable loading when parsing the `feature_store.yaml` file, or optionally make it possible to set values inside the file using env vars. That way we are keeping track of all options in a single place, and loading of env vars happens during initialization, not in a place deeply nested in the code.

abrookins replied:
Ah, that's a good idea! I agree that it's best to have a single place that all of the configuration options are visible.

abrookins commented on sdk/python/feast/infra/redis_provider.py:
```diff
@@ -0,0 +1,276 @@
+            redis_key_bin = _redis_key(
+                project, EntityKeyProto(join_keys=table_join_keys)
+            )
+            keys = client.keys(f"{redis_key_bin}*")
```
This is a general Redis tip, but you should never use the `KEYS` command in production because it loads all matching keys into memory. That can block the server and use too much memory. Instead, use the `SCAN` family of commands. In this case, you can use redis-py's `scan_iter()` method -- something like:

```
keys = [k for k in redis.scan_iter(match=f'{redis_key_bin}*', count=100)]
if keys:
    client.unlink(*keys)
```

woop replied:
Thanks @abrookins, good catch!

qooba replied:
Thanks @abrooking I have fixed this :)

abrookins replied:
Awesome! 👍

abrookins reviewed (changes requested):
I only found one issue, which is the use of `KEYS` instead of `SCAN`.

Also, @qooba, do you want me to test this out manually? If so, could you send whatever project setup you used for testing?

qooba commented:
@woop following your suggestion I have moved ```redis_type``` to the ```feature_store.yaml``` configuration. Additionally ```redis_connection_string``` will be filled using ```REDIS_CONNECTION_STRING``` environment variable during configuration parsing.

abrookins reviewed (approved):
Looks great - nice work! 👏

woop commented:
> Looks great - nice work!

Thanks for the review @abrookins!

woop commented on .github/workflows/pr_integration_tests.yml:
```diff
@@ -67,3 +67,4 @@ jobs:
           env_vars: OS,PYTHON
           fail_ci_if_error: true
           verbose: true
+
```
Can you remind me why we don't test Redis for PRs?

woop commented on sdk/python/feast/infra/redis_provider.py:
```diff
@@ -0,0 +1,281 @@
+import json
```
`redis.py` for the file name is fine, for consistency.

woop commented on sdk/python/feast/infra/redis_provider.py:
```diff
@@ -0,0 +1,281 @@
+
+    def __init__(self, config: RepoConfig):
+        assert isinstance(config.online_store, RedisOnlineStoreConfig)
+        if config and config.online_store:
```
It's not clear why there is a conditional here if `isinstance` implies that this config exists. I think you should assume that `config.online_store` always exists, and we should provide defaults to these options in all cases. So you should just reach for the config without testing whether they exist.

abrookins reviewed (commented):
I tried to walk through the Feast Quickstart using the Redis provider but ran into a few issues.

First, this assertion in repo_config.py never passes -- it seems like the list is missing "redis":

```python
       # Make sure the user hasn't provided the wrong type
       assert online_store_type in ["datastore", "sqlite"]
```

Second, the Redis provider tries to import a helper function that doesn't exist:

```python
from feast.infra.offline_stores.helpers import get_offline_store_from_sources
```

The function `get_offline_store_from_sources` doesn't exist in the `helpers` module, just `get_offline_store_from_sources()`. @qooba, did you leave the `get_offline_store_from_sources` helper out of the commit by accident?

qooba commented:
@abrookins I have missed changes after rebase, now it is fixes. 
You can use feature_store.yaml
```
project: default
registry: data/registry.db
provider: redis
online_store:
  redis_type: redis # or redis_cluster
```

you define redis connection string in environment variable  eg.
```
%env REDIS_CONNECTION_STRING=redis:6379,db=0
```

@woop : 
* ```redis_provider.py``` -> ```redis.py``` done
* unnecessary if remove done
* ```pr_integration.yml``` <-> ```integration.yml``` done 
* ```redist_type``` - moved to ```feature_store.yaml``` (to ```RedisOnlineStoreConfig```)
* ```redis_connection_string``` - removed from provider and moved to ```RedisOnlineStoreConfig``` but still it will be taken from ```REDIS_CONNECTION_STRING``` environment variable rather than from ```feature_store.yaml```

qooba commented:
@abrookins during local tests 
* I have used the parquets generated by ```feast init``` I have only changed the provider and added 
```env REDIS_CONNECTION_STRING=redis:6379,db=0```.
* I have setup the feast ver 0.9 using helm go through ```minimal_ride_hailing.ipynb``` and then I have used feast ver 0.10 to fetch online features materialized by ver 0.9

abrookins commented:
Looking a lot better! However, I still can't complete the Quickstart. Here are the commands I ran and their output:

```
env ❯ cat -p feature_store.yaml 
project: default
registry: data/registry.db
provider: redis
online_store:
  redis_type: redis 
  
env ❯ feast apply
Registered entity driver_id
Registered feature view driver_hourly_stats
Deploying infrastructure for driver_hourly_stats

(Run Jupyter notebook code in the example from Quickstart: https://docs.feast.dev/quickstart)

env ❯ CURRENT_TIME=$(date -u +"%Y-%m-%dT%H:%M:%S")
feast materialize-incremental $CURRENT_TIME
Materializing 1 feature views to 2021-05-26 22:17:13-07:00 into the redis online store.

driver_hourly_stats from 2021-05-25 22:17:14-07:00 to 2021-05-26 22:17:13-07:00:
0it [00:00, ?it/s]
```

At this point, Redis is empty. Running `MONITOR` while materializing no commands run, which seems odd.

Then if I try to use a Notebook, I don't get any data from the online store (not surprising due to the above):

```
from pprint import pprint
from feast import FeatureStore

store = FeatureStore(repo_path=".")

feature_vector = store.get_online_features(
    feature_refs=[
        'driver_hourly_stats:conv_rate',
        'driver_hourly_stats:acc_rate',
        'driver_hourly_stats:avg_daily_trips'
    ],
    entity_rows=[{"driver_id": 1001}]
).to_dict()

pprint(feature_vector)

{'driver_hourly_stats__acc_rate': [None],
 'driver_hourly_stats__avg_daily_trips': [None],
 'driver_hourly_stats__conv_rate': [None],
 'driver_id': [1001]}
```

Are you able to work through the Quickstart with this provider? Maybe I'm missing a step.

qooba commented:
@abrookins - I have also noticed this before thus I use:
```feast materialize 2020-04-26T22:05:26 2021-06-26T22:05:26```
I still have to check if materialize-incremental does not work for redis provider or if there is a problem for all providers.

abrookins commented:
Interesting! I tried the same steps with the local provider, and the result
is the same -- nothing materializes. So, I think the steps in the
Quickstart must not work with 0.10.

The good news: `feast materialize 2020-04-26T22:05:26 2021-06-26T22:05:26`
worked with your new Redis provider!




On Wed, May 26, 2021 at 3:34 PM Kuba Sołtys ***@***.***>
wrote:

> @abrookins <https://github.com/abrookins> - I have also noticed this
> before thus I use:
> feast materialize 2020-04-26T22:05:26 2021-06-26T22:05:26
> I still have to check if materialize-incremental does not work for redis
> provider or if there is a problem for all providers.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/feast-dev/feast/pull/1511#issuecomment-849164717>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAXXHWNRMBYAAES2DMN7NTTPVZNPANCNFSM43VXZOTA>
> .
>

Disclaimer

The information contained in this communication from the sender is confidential. It is intended solely for use by the recipient and others authorized to receive it. If you are not the recipient, you are hereby notified that any disclosure, copying, distribution or taking action in relation of the contents of this information is strictly prohibited and may be unlawful.

This email has been scanned for viruses and malware, and may have been automatically archived by Mimecast Ltd, an innovator in Software as a Service (SaaS) for business. Providing a safer and more useful place for your human generated data. Specializing in; Security, archiving and compliance. To find out more visit the Mimecast website.

woop commented:
> Interesting! I tried the same steps with the local provider, and the result is the same -- nothing materializes. So, I think the steps in the Quickstart must not work with 0.10. The good news: `feast materialize 2020-04-26T22:05:26 2021-06-26T22:05:26` worked with your new Redis provider!

Creating a card to track https://github.com/feast-dev/feast/issues/1588

qooba commented on sdk/python/feast/infra/redis_provider.py:
```diff
@@ -0,0 +1,281 @@
+        feature_view.materialization_intervals.append((start_date, end_date))
+        registry.apply_feature_view(feature_view, project)
+
+    def _get_cs(self):
```
changed to ```_parse_connection_string```

qooba commented on sdk/python/feast/infra/redis_provider.py:
```diff
@@ -0,0 +1,281 @@
+        registry: Registry,
+        project: str,
+    ) -> RetrievalJob:
+        offline_store = get_offline_store_from_sources(
```
fixed

qooba commented on sdk/python/feast/infra/redis_provider.py:
```diff
@@ -0,0 +1,281 @@
+        entity_names=entity_key.join_keys,
+        entity_values=entity_key.entity_values,
+    )
+    # key = _mmh3(serialize_entity_key(entity_key))
```
fixed :)

qooba commented on sdk/python/feast/repo_config.py:
```diff
@@ -128,6 +151,13 @@ def _validate_online_store_config(cls, values):
                 values["online_store"]["type"] = "datastore"
+            elif values["provider"] == "redis":
+                values["online_store"]["type"] = "redis"
+                values["online_store"]["redis_connection_string"] = os.environ.get(
```
@woop the connection string will contain password thus the ```feature_store.yaml``` is not the best place to keep secrets. In most cases the redis password should be kept in vault or in k8s secrets. The environment variable is the simplest way to share it with Feast, another solution is to set the value after ```FeatureStore``` instantiation (but I think it will be less intuitive for users).

woop replied:
While I agree with that, I also think that we are adding complexity here that we can offload to the user for the time being. I think we'll need to have a separate PR by somebody in order to add env var handling to Feast. Right now, teams can still template their configuration file with the secret in production (which is actually safer than env vars since they are unlikely to be accidentally printed out) or they can fill in the secret themselves using a script during deployment.

What do you think? Can we delay this functionality for now?

qooba commented on sdk/python/setup.py:
```diff
@@ -56,6 +56,7 @@
     "tabulate==0.8.*",
     "toml==0.10.*",
     "tqdm==4.*",
+    "redis-py-cluster==2.1.2",
```
done

woop commented:
@qooba this is fantastic work!

We really appreciate the effort you have put in here. I'm planning to do my final checks on this PR early next week, but I don't think there is anything outstanding any more. I just want to actually run it myself and make sure everything works as expected. Thank you for being so patient :pray:

abrookins commented:
Tips for testing:

Feature_store.yaml:
```
project: polished_guinea
registry: data/registry.db
provider: redis
online_store:
  type: redis
  redis_type: redis
  redis_connection_string: "localhost:6379"
```

And make sure you install redis-py-cluster, not rediscluster — rediscluster is not maintained AFAIK.

qooba commented:
Thanks @abrookins . Doublechecked I'm using ```redis-py-cluster``` :)

woop commented:
Ideally we'd have compatibility between the offline and online store implementations, but there isn't a guarantee. We are using providers as the top-level abstraction to ensure that there is always an escape hatch for ensuring compatibility.

For what it's worth, we're going to pull out OnlineStore into its own class https://docs.google.com/document/d/1MD0aS2_hGzd1tJ7DNjE3NgEtcuekh3O06OeQ9aavylY/edit#

abrookins commented:
@qooba @woop Thanks for the context. I have a follow-up question, if I may! My reading of ` _validate_offline_store_config()` suggests that nothing, at least in that method, would stop you from configuring  BigQuery for offline storage and Redis for online storage — but that if you configured an online provider, and it was Redis, and did not configure an offline provider, then you’d get file as the default for Redis.

Am I missing something, then, or could you in fact use BigQuery for offline and Redis for online, even if there is no guarantee about compatibility?

woop commented:
> @qooba @woop Thanks for the context. I have a follow-up question, if I may! My reading of ` _validate_offline_store_config()` suggests that nothing, at least in that method, would stop you from configuring BigQuery for offline storage and Redis for online storage — but that if you configured an online provider, and it was Redis, and did not configure an offline provider, then you’d get file as the default for Redis.
> 
> Am I missing something, then, or could you in fact use BigQuery for offline and Redis for online, even if there is no guarantee about compatibility?

Thanks for paying so close attention @abrookins.

The idea is that we would NOT have Redis as a provider. Instead, we would have only two providers for now (`local` and `gcp`). As soon as this PR lands, we will refactor it using #1628. That would allow `local` and `gcp` providers to opt into supporting `Redis` with `BigQuery` and/or `File` sources. Ultimately though, the provider is the one that decides. 

Redis itself will not be a provider.

woop commented on sdk/python/feast/repo_config.py:
```diff
@@ -52,7 +53,28 @@ class DatastoreOnlineStoreConfig(FeastBaseModel):
+    redis_type: RedisType = RedisType.redis
+    """Redis type: redis or redis_cluster"""
+
+    redis_connection_string: StrictStr
```
we can just make this `connection_string` right?

woop commented on sdk/python/feast/repo_config.py:
```diff
@@ -52,7 +53,28 @@ class DatastoreOnlineStoreConfig(FeastBaseModel):
+    type: Literal["redis"] = "redis"
+    """Online store type selector"""
+
+    redis_type: RedisType = RedisType.redis
```
should we default to non-cluster mode?

woop replied:
Nevermind, I was getting a redis-cluster exception. It was due to a dependency conflict.

woop commented:
/lgtm

abrookins commented:
> > @qooba @woop Thanks for the context. I have a follow-up question, if I may! My reading of ` _validate_offline_store_config()` suggests that nothing, at least in that method, would stop you from configuring BigQuery for offline storage and Redis for online storage — but that if you configured an online provider, and it was Redis, and did not configure an offline provider, then you’d get file as the default for Redis.
> > Am I missing something, then, or could you in fact use BigQuery for offline and Redis for online, even if there is no guarantee about compatibility?
> 
> Thanks for paying so close attention @abrookins.
> 
> The idea is that we would NOT have Redis as a provider. Instead, we would have only two providers for now (`local` and `gcp`). As soon as this PR lands, we will refactor it using #1628. That would allow `local` and `gcp` providers to opt into supporting `Redis` with `BigQuery` and/or `File` sources. Ultimately though, the provider is the one that decides.
> 
> Redis itself will not be a provider.

Ahh, thank you for clarifying. I thought I might not have had the language quite right!

woop commented:
/lgtm
